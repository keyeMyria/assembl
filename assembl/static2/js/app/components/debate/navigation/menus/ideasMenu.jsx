// @flow
import React from 'react';
import { connect } from 'react-redux';
import { graphql, compose } from 'react-apollo';

import AllIdeasQuery from '../../../../graphql/AllIdeasQuery.graphql';
import withLoadingIndicator from '../../../common/withLoadingIndicator';
import MenuList, { type ItemNode } from '.';

type IdeasTableProps = {
  identifier: string,
  kind: string,
  onMenuItemClick: Function,
  data: {
    loading: boolean,
    error: Object,
    ideas: Array<ItemNode>,
    rootIdea: ItemNode
  }
};

export function DumbIdeasMenu(props: IdeasTableProps) {
  const { identifier, kind, onMenuItemClick, data } = props;
  const { ideas, rootIdea } = data;
  return (
    <MenuList
      items={ideas}
      rootItem={rootIdea && rootIdea.id}
      identifier={identifier}
      onMenuItemClick={onMenuItemClick}
      kind={kind}
    />
  );
}

const IdeasTableWithData = graphql(AllIdeasQuery);

const mapStateToProps = state => ({
  lang: state.i18n.locale,
  debate: state.debate
});

export default compose(connect(mapStateToProps), IdeasTableWithData, withLoadingIndicator())(DumbIdeasMenu);