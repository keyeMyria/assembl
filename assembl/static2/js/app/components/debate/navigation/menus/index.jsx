// @flow
import React from 'react';
import TabMenuList from './tableMenuList';
import IndexMenuList from './indexMenuList';

export const MENU_KINDS = {
  index: 'index',
  table: 'table'
};

export type ItemNode = {
  id: string,
  title: string,
  img: {
    externalUrl: string
  },
  numContributors: number,
  numPosts: number,
  parentId: string
};

type MenuListProps = {
  kind: string
};

function MenuList(props: MenuListProps) {
  switch (props.kind) {
  case MENU_KINDS.index:
    return <IndexMenuList {...props} />;
  default:
    return <TabMenuList {...props} />;
  }
}

export { default as SurveyMenu } from './surveyMenu';
export { default as IdeasMenu } from './ideasMenu';
export { default as Menu } from './menu';
export default MenuList;