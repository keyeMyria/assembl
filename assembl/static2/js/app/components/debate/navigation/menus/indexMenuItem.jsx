// @flow
import React from 'react';
import { Link } from 'react-router';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { get } from '../../../../utils/routeMap';
import IdeasIndex from '../../common/ideasIndex';
import { SECTION_INDEX_GENERATOR } from '../../../../utils/section';
import { getChildrenFromArray } from '../../../../utils/tree';
import { type ItemNode } from '.';

type MenuItemProps = {
  index: number,
  items: Array<ItemNode>,
  item: ItemNode,
  identifier: string,
  selected: boolean,
  slug: string,
  indexGenerator: Function,
  onChange: Function
};

export class DumbIndexMenuItem extends React.Component<*, MenuItemProps, void> {
  static defaultProps = {
    index: 1,
    indexGenerator: SECTION_INDEX_GENERATOR.alphanumericOr,
    parents: []
  };

  arrow = null;

  toggleMenu = (event: SyntheticMouseEvent<>) => {
    const { onChange, item, selected } = this.props;
    // scroll to the item
    const arrow = this.arrow;
    const menuContainer = document.getElementById('navbar-burger');
    if (menuContainer && arrow) {
      menuContainer.scrollTop = arrow.offsetTop;
    }
    if (onChange) onChange(!selected, item.id);
    event.preventDefault();
  };

  getTitle = () => {
    const { item, indexGenerator, index } = this.props;
    return `${indexGenerator([index])} ${item.title}`;
  };

  render() {
    const { index, identifier, item, selected, items, slug } = this.props;
    const { id, title } = item;
    const subItems = getChildrenFromArray(id, items);
    return (
      <div className="index-menu-item-container">
        <Link className="index-item" to={`${get('themeInPhase', { slug: slug, phase: identifier, themeId: id })}`}>
          <div className="thumb-body">
            <div title={title} className="thumb-title">
              {this.getTitle()}
            </div>
          </div>
          {subItems.length > 0 && (
            <span
              ref={(arrow) => {
                this.arrow = arrow;
              }}
              onClick={this.toggleMenu}
              className={classNames('thumb-arrow assembl-icon', {
                'assembl-icon-down-open': !selected,
                'assembl-icon-up-open': selected
              })}
            />
          )}
        </Link>
        <IdeasIndex
          parents={[index]}
          className={classNames('sub-index sub-index-1', { closed: !selected })}
          items={items}
          rootItem={id}
          identifier={identifier}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  slug: state.debate.debateData.slug
});

export default connect(mapStateToProps)(DumbIndexMenuItem);