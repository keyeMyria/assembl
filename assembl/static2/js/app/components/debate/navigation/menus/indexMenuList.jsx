// @flow
import React from 'react';

import IndexMenuItem from './indexMenuItem';
import { getChildrenFromArray } from '../../../../utils/tree';
import { type ItemNode } from '.';

type MenuListProps = {
  items: Array<ItemNode>,
  rootItem: string,
  identifier: string,
  className: string
};

type MenuListState = {
  selectedItem: string
};

class IndexMenuList extends React.Component<*, MenuListProps, MenuListState> {
  props: MenuListProps;

  static defaultProps = {
    className: '',
    parents: []
  };

  state = {
    selectedItem: ''
  };

  componentWillReceiveProps() {
    if (this.state.selectedItem) {
      this.setState({ selectedItem: '' });
    }
  }

  onItemChange = (active: boolean, itemId: string) => {
    this.setState({ selectedItem: active ? itemId : '' });
  };

  render() {
    const { rootItem, items, identifier, className } = this.props;
    const { selectedItem } = this.state;
    const rootItems = getChildrenFromArray(rootItem, items);
    if (rootItems.length === 0) return null;
    return (
      <div className={className}>
        {rootItems.map((item, index) => (
          <IndexMenuItem
            key={item.id}
            item={item}
            index={index + 1}
            items={items}
            selected={item.id === selectedItem}
            onChange={this.onItemChange}
            identifier={identifier}
          />
        ))}
      </div>
    );
  }
}

export default IndexMenuList;