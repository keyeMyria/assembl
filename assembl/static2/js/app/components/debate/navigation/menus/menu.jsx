// @flow
import React from 'react';
import type { ApolloClient } from 'react-apollo';

import { SurveyMenu, IdeasMenu } from '.';
import AllIdeasQuery from '../../../../graphql/AllIdeasQuery.graphql';
import DebateThematicsQuery from '../../../../graphql/DebateThematicsQuery.graphql';

const queries = {
  survey: DebateThematicsQuery,
  default: AllIdeasQuery
};

export function prefetchMenuQuery(client: ApolloClient, variables: Object) {
  const query = queries[variables.identifier];
  client.query({
    query: query || queries.default,
    variables: variables
  });
}

type MenuTableProps = {
  identifier: string
};

function Menu(props: MenuTableProps) {
  switch (props.identifier) {
  case 'survey':
    return <SurveyMenu {...props} />;
  default:
    return <IdeasMenu {...props} />;
  }
}

export default Menu;