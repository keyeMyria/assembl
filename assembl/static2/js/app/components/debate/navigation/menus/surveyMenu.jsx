// @flow
import React from 'react';
import { connect } from 'react-redux';
import { graphql, compose } from 'react-apollo';
import DebateThematicsQuery from '../../../../graphql/DebateThematicsQuery.graphql';
import withLoadingIndicator from '../../../common/withLoadingIndicator';
import MenuList, { type ItemNode } from '.';

type SurveyTableProps = {
  identifier: string,
  kind: string,
  onMenuItemClick: Function,
  data: {
    loading: boolean,
    error: Object,
    thematics: Array<ItemNode>
  }
};

export function DumbSurveyMenu(props: SurveyTableProps) {
  const { identifier, kind, onMenuItemClick, data } = props;
  const { thematics } = data;
  return <MenuList items={thematics} identifier={identifier} onMenuItemClick={onMenuItemClick} kind={kind} />;
}

const SurveyTableWithData = graphql(DebateThematicsQuery);

const mapStateToProps = state => ({
  lang: state.i18n.locale,
  debate: state.debate
});

export default compose(connect(mapStateToProps), SurveyTableWithData, withLoadingIndicator())(DumbSurveyMenu);