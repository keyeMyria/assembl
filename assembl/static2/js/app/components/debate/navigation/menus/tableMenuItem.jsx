// @flow
import React from 'react';
import { Link } from 'react-router';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { get } from '../../../../utils/routeMap';
import PostsAndContributorsCount from '../../../common/postsAndContributorsCount';
import { type ItemNode } from '.';

type MenuItemProps = {
  item: ItemNode,
  identifier: string,
  selected: boolean,
  hasSubItems: boolean,
  slug: string,
  onClick: Function,
  onMouseOver: Function,
  onMouseLeave: Function
};

type MenuItemState = {
  active: boolean
};

export class DumbTableMenuItem extends React.Component<*, MenuItemProps, MenuItemState> {
  state = {
    active: false
  };

  showMenu = () => {
    const { onMouseOver, item } = this.props;
    this.setState({ active: true }, () => {
      if (onMouseOver) onMouseOver(item.id);
    });
  };

  hideMenu = () => {
    const { onMouseLeave, item } = this.props;
    this.setState({ active: false }, () => {
      if (onMouseLeave) onMouseLeave(item.id);
    });
  };

  render() {
    const { identifier, item, selected, hasSubItems, slug, onClick } = this.props;
    const { id, title, img, numContributors, numPosts } = item;
    const { active } = this.state;
    const isSelected = active || selected;
    const displayArrow = isSelected && hasSubItems;
    return (
      <div
        className={classNames('menu-item-container', {
          active: isSelected,
          empty: isSelected && !displayArrow
        })}
        onMouseOver={this.showMenu}
        onMouseLeave={this.hideMenu}
      >
        <Link
          onClick={onClick}
          className="menu-item"
          to={`${get('themeInPhase', { slug: slug, phase: identifier, themeId: id })}`}
        >
          <div className="thumb-img" style={img && img.externalUrl ? { backgroundImage: `url(${img.externalUrl})` } : null}>
            <div className="thumb-img-background" />
          </div>
          <div className="thumb-body">
            <div title={title} className="thumb-title">
              {title}
            </div>
            <PostsAndContributorsCount className="menu-stats" numContributors={numContributors} numPosts={numPosts} />
          </div>
          {displayArrow && <span className="thumb-arrow assembl-icon assembl-icon-right-dir" />}
        </Link>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  slug: state.debate.debateData.slug
});

export default connect(mapStateToProps)(DumbTableMenuItem);