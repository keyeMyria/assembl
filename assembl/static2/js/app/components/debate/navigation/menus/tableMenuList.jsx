// @flow
import React from 'react';
import classNames from 'classnames';

import TableMenuItem from './tableMenuItem';
import { type ItemNode } from '.';

type MenuListProps = {
  items: Array<ItemNode>,
  rootItem: string,
  identifier: string,
  className: string,
  onMenuItemClick: Function
};

type MenuListState = {
  selected: string
};

class TableMenuList extends React.Component<*, MenuListProps, MenuListState> {
  props: MenuListProps;

  static defaultProps = {
    className: ''
  };

  state = {
    selected: ''
  };

  componentWillReceiveProps() {
    if (this.state.selected) {
      this.setState({ selected: '' });
    }
  }

  onItemOver = (itemId: string) => {
    this.setState({ selected: itemId });
  };

  getItemChildren = (itemId: string) => {
    const { items } = this.props;
    if (items) {
      return itemId ? items.filter(item => item.parentId === itemId) : items;
    }
    return [];
  };

  render() {
    const { items, rootItem, identifier, className, onMenuItemClick } = this.props;
    const { selected } = this.state;
    const rootItems = this.getItemChildren(rootItem);
    if (rootItems.length === 0) return null;
    return (
      <div className={classNames('menu-table-col', className)}>
        <div className="menu-table">
          {rootItems.map(item => (
            <TableMenuItem
              key={item.id}
              hasSubItems={items.some(listItem => listItem.parentId === item.id)}
              selected={item.id === selected}
              onClick={onMenuItemClick}
              onMouseOver={this.onItemOver}
              identifier={identifier}
              item={item}
            />
          ))}
        </div>
        {selected && (
          <TableMenuList
            onMenuItemClick={onMenuItemClick}
            className="sub-menu"
            items={items}
            rootItem={selected}
            identifier={identifier}
          />
        )}
      </div>
    );
  }
}

export default TableMenuList;