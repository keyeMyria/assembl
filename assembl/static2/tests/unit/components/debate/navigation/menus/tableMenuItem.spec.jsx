import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbTableMenuItem } from '../../../../../../js/app/components/debate/navigation/menus/tableMenuItem';

describe('DumbMenuItem component', () => {
  it('should match a selected menu item', () => {
    const props = {
      item: {
        id: 'fooId',
        title: 'Foo',
        img: {
          externalUrl: 'https://foo.bar/img'
        },
        numContributors: 10,
        numPosts: 123
      },
      identifier: 'survey',
      selected: true,
      hasSubItems: true,
      slug: 'slug'
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbTableMenuItem {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  it('should match a not selected menu item', () => {
    const props = {
      item: {
        id: 'fooId',
        title: 'Foo',
        img: {
          externalUrl: 'https://foo.bar/img'
        },
        numContributors: 10,
        numPosts: 123
      },
      identifier: 'survey',
      selected: false,
      hasSubItems: true,
      slug: 'slug'
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbTableMenuItem {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});