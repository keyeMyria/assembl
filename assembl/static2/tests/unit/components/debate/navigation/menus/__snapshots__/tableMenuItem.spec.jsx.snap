// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbMenuItem component should match a not selected menu item 1`] = `
<div
  className="menu-item-container"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <Link
    className="menu-item"
    onClick={undefined}
    onlyActiveOnIndex={false}
    style={Object {}}
    to="/slug/debate/survey/theme/fooId"
  >
    <div
      className="thumb-img"
      style={
        Object {
          "backgroundImage": "url(https://foo.bar/img)",
        }
      }
    >
      <div
        className="thumb-img-background"
      />
    </div>
    <div
      className="thumb-body"
    >
      <div
        className="thumb-title"
        title="Foo"
      >
        Foo
      </div>
      <PostsAndContributorsCount
        className="menu-stats"
        numContributors={10}
        numPosts={123}
      />
    </div>
  </Link>
</div>
`;

exports[`DumbMenuItem component should match a selected menu item 1`] = `
<div
  className="menu-item-container active"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <Link
    className="menu-item"
    onClick={undefined}
    onlyActiveOnIndex={false}
    style={Object {}}
    to="/slug/debate/survey/theme/fooId"
  >
    <div
      className="thumb-img"
      style={
        Object {
          "backgroundImage": "url(https://foo.bar/img)",
        }
      }
    >
      <div
        className="thumb-img-background"
      />
    </div>
    <div
      className="thumb-body"
    >
      <div
        className="thumb-title"
        title="Foo"
      >
        Foo
      </div>
      <PostsAndContributorsCount
        className="menu-stats"
        numContributors={10}
        numPosts={123}
      />
    </div>
    <span
      className="thumb-arrow assembl-icon assembl-icon-right-dir"
    />
  </Link>
</div>
`;
